File: sdks/rust/src/nodes/api_call.rs
use async_trait::async_trait;
use std::collections::HashMap;

use crate::node::NodeHandler;
use crate::types::Context;

/// ApiCallNode makes HTTP requests to external APIs.
///
/// Config:
///   - `url` (string, required): The URL to call
///   - `method` (string, optional): HTTP method (default: GET)
///   - `timeout` (number, optional): Timeout in seconds (default: 10)
pub struct ApiCallNode;

type BoxError = Box<dyn std::error::Error + Send + Sync>;

#[async_trait]
impl NodeHandler for ApiCallNode {
    async fn execute(
        &self,
        ctx: &mut Context,
        config: &HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, BoxError> {
        let url = config
            .get("url")
            .and_then(|v| v.as_str())
            .ok_or_else(|| -> BoxError {
                "'url' is required in node config".into()
            })?;

        let method = config
            .get("method")
            .and_then(|v| v.as_str())
            .unwrap_or("GET")
            .to_uppercase();

        let timeout = config
            .get("timeout")
            .and_then(|v| v.as_u64())
            .unwrap_or(10);

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(timeout))
            .build()?;

        let mut req_builder = match method.as_str() {
            "POST" => client.post(url),
            "PUT" => client.put(url),
            "PATCH" => client.patch(url),
            "DELETE" => client.delete(url),
            "HEAD" => client.head(url),
            _ => client.get(url),
        };

        // Add request body for POST/PUT/PATCH
        if matches!(method.as_str(), "POST" | "PUT" | "PATCH") {
            if let Some(body) = ctx.request.body.get("body") {
                req_builder = req_builder.json(body);
            }
        }

        // Add headers from config
        if let Some(headers) = config.get("headers").and_then(|v| v.as_object()) {
            for (k, v) in headers {
                if let Some(val) = v.as_str() {
                    req_builder = req_builder.header(k.as_str(), val);
                }
            }
        }

        let response: reqwest::Response = req_builder.send().await?;

        let status = response.status().as_u16();
        let resp_headers: HashMap<String, String> = response
            .headers()
            .iter()
            .map(|(k, v): (&reqwest::header::HeaderName, &reqwest::header::HeaderValue)| {
                (k.to_string(), v.to_str().unwrap_or("").to_string())
            })
            .collect();

        let body_text: String = response.text().await?;

        let data: serde_json::Value =
            serde_json::from_str(&body_text).unwrap_or(serde_json::Value::String(body_text));

        Ok(serde_json::json!({
            "status": status,
            "data": data,
            "headers": resp_headers,
        }))
    }
}

File: sdks/go/examples/nodes/apicall.go
package nodes

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	blok "github.com/nickincloud/blok-go"
)

// ApiCallNode makes HTTP requests to external APIs.
//
// Config:
//   - url (string, required): The URL to call
//   - method (string, optional): HTTP method (default: GET)
//   - headers (map[string]string, optional): Additional headers
//   - timeout (number, optional): Timeout in seconds (default: 10)
//
// Input (request body):
//   - body (any, optional): Request body for POST/PUT/PATCH
//
// Output:
//   - status (number): HTTP status code
//   - data (any): Parsed response body
//   - headers (map[string]string): Response headers
type ApiCallNode struct{}

// Execute performs the HTTP request.
func (n *ApiCallNode) Execute(ctx *blok.Context, config map[string]interface{}) (interface{}, error) {
	// Get URL from config (required)
	url, ok := config["url"].(string)
	if !ok || url == "" {
		return nil, blok.NewConfigurationError("'url' is required in node config")
	}

	// Get method from config, default to GET
	method := "GET"
	if v, ok := config["method"].(string); ok && v != "" {
		method = strings.ToUpper(v)
	}

	// Get timeout from config, default to 10 seconds
	timeout := 10
	if v, ok := config["timeout"].(float64); ok && v > 0 {
		timeout = int(v)
	}

	// Prepare request body
	var reqBody io.Reader
	if body := ctx.Request.BodyMap(); body != nil {
		if bodyData, exists := body["body"]; exists && bodyData != nil {
			jsonData, err := json.Marshal(bodyData)
			if err != nil {
				return nil, blok.NewExecutionError("failed to marshal request body", err)
			}
			reqBody = bytes.NewBuffer(jsonData)
		}
	}

	// Create HTTP request
	req, err := http.NewRequest(method, url, reqBody)
	if err != nil {
		return nil, blok.NewNetworkError(fmt.Sprintf("failed to create request to %s", url), err)
	}

	// Set headers from config
	if headers, ok := config["headers"].(map[string]interface{}); ok {
		for k, v := range headers {
			if s, ok := v.(string); ok {
				req.Header.Set(k, s)
			}
		}
	}

	if reqBody != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	// Execute request
	client := &http.Client{
		Timeout: time.Duration(timeout) * time.Second,
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, blok.NewNetworkError(fmt.Sprintf("request to %s failed", url), err)
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, blok.NewNetworkError("failed to read response body", err)
	}

	// Try to parse as JSON, fall back to string
	var responseData interface{}
	if err := json.Unmarshal(respBody, &responseData); err != nil {
		responseData = string(respBody)
	}

	// Collect response headers
	respHeaders := make(map[string]string)
	for k := range resp.Header {
		respHeaders[k] = resp.Header.Get(k)
	}

	return map[string]interface{}{
		"status":  resp.StatusCode,
		"data":    responseData,
		"headers": respHeaders,
	}, nil
}
